arches/app/views/api/controlled_lists.py

```diff
@@ -0,0 +1,413 @@
+)
+class ControlledListsView(View):
+    def get(self, request):
+        """Returns either a flat representation (?flat=true) or a tree (default)."""
+        lists = (
+            ControlledList.objects.annotate_node_fields(
+                node_ids="pk",
+                node_names="name",
+                nodegroup_ids="nodegroup_id",
+                graph_ids="graph_id",
+                graph_names="graph__name",
+            )
+            .order_by("name")
+            .prefetch_related(*_prefetch_terms(request))
+        )
+
+        flat = str_to_bool(request.GET.get("flat", "false"))
+        permitted = [
+            ng.pk for ng in get_nodegroups_by_perm(request.user, "read_nodegroup")
+        ]
+        serialized = [
+            obj.serialize(flat=flat, permitted_nodegroups=permitted) for obj in lists
+        ]
+
+        return JSONResponse({"controlled_lists": serialized})
+
+
+@method_decorator(
+    group_required("RDM Administrator", raise_exception=True), name="dispatch"
+)
+class ControlledListView(View):
+    def get(self, request, **kwargs):
+        """Returns either a flat representation (?flat=true) or a tree (default)."""
+        list_id = kwargs.get("id")
+        try:
+            lst = ControlledList.objects.prefetch_related(
+                *_prefetch_terms(request)
+            ).get(pk=list_id)
+        except ControlledList.DoesNotExist:
+            return JSONErrorResponse(status=HTTPStatus.NOT_FOUND)
+
+        flat = str_to_bool(request.GET.get("flat", "false"))
+        permitted = [
+            ng.pk for ng in get_nodegroups_by_perm(request.user, "read_nodegroup")
+        ]
+        serialized = lst.serialize(flat=flat, permitted_nodegroups=permitted)
+
+        return JSONResponse(serialized)
+
+    def post(self, request, **kwargs):
+        data = JSONDeserializer().deserialize(request.body)
+
+        # Update
+        if "id" in kwargs:
+            raise NotImplementedError
+
+        # Insert
+        name = data.get("name", None)
+        lst = ControlledList(name=name)
+        try:
+            lst.full_clean()  # applies default name
+        except ValidationError as ve:
+            return JSONErrorResponse(
+                message="\n".join(ve.messages), status=HTTPStatus.BAD_REQUEST
+            )
+        lst.save()
+        return JSONResponse(lst.serialize(), status=HTTPStatus.CREATED)
+
+    def patch(self, request, **kwargs):
+        list_id: UUID = kwargs.get("id")
+        data = JSONDeserializer().deserialize(request.body)
+        data.pop("items", None)
+        sortorder_map = data.pop("sortorder_map", {})
+        parent_map = data.pop("parent_map", {})
+
+        update_fields = list(data)
+        if not update_fields and not sortorder_map:
+            return JSONResponse(status=HTTPStatus.BAD_REQUEST)
+
+        clist = ControlledList(id=list_id, **data)
+
+        exclude_fields = {f for f in field_names(clist) if f not in update_fields}
```

Comment:
nit: single-letter naming. Not a fan of the other short names throughout but not gonna block because of them.

Reply:
For me, this is the same situation with comprehension variables we discussed in https://github.com/archesproject/arches/pull/11016#discussion_r1636476703.

But I can probably find something agreeable to both of us by backing out of this entirely in favor of set difference.

Reply:
ah yeah I hear you and I'm open to shorter naming like we discussed for sure 👍 . That said I'll never approve single-letter variable names outside of a lambda 😉